src/NHibernate/Type/AbstractDateTimeType.cs

```diff
@@ -91,8 +92,38 @@ protected virtual DateTime GetDateTime(DbDataReader rs, int index, ISessionImple
 		/// <param name="value">The original value.</param>
 		/// <param name="session">The session for which the operation is done.</param>
 		/// <returns>The <see cref="DateTime" /> to set as parameter value.</returns>
-		protected virtual DateTime GetDateTimeToSet(object value, ISessionImplementor session) =>
-			AdjustDateTime((DateTime) value);
+		protected virtual DateTime GetDateTimeToSet(object value, ISessionImplementor session)
+		{
+			var dateValue = (DateTime) value;
+			if (Kind == DateTimeKind.Unspecified)
+				return AdjustDateTime(dateValue);
+			if (dateValue.Kind == DateTimeKind.Unspecified)
+				throw new ArgumentException($"Date kind is unspecified", nameof(value));
+			if (Kind != dateValue.Kind)
+			{
+				if (Kind == DateTimeKind.Utc)
+				{
+					if (TimeZoneInfo.Local.IsInvalidTime(dateValue))
+						throw new ArgumentException(
+							$"Date is local and has an invalid time ({dateValue:o}), unable to convert to UTC.",
+							nameof(value));
+					if (TimeZoneInfo.Local.IsAmbiguousTime(dateValue))
+						throw new ArgumentException(
+							$"Date is local and has an ambiguous time ({dateValue:o}), unable to convert to UTC.",
+							nameof(value));
```

Comment:
Conversion from Local to Utc may cause issue when daylight savings change hour, going back one hour. Having code failing one hour per year is not really nice and easy to test, so maybe should we just not support converting from local to utc, and only support converting from utc to local (which can always be done). Or not support any conversion at all.

Reply:
👍 for not doing any conversion